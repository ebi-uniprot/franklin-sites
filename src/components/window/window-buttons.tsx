import { ReactNode } from 'react';
import cn from 'classnames';

import CloseIcon from '../../svg/times.svg';

type WindowActionButtonProps = {
  icon?: ReactNode;
  text?: string;
  onClick: () => void;
  primary?: boolean;
  className?: string;
};
export const WindowActionButton = ({
  icon,
  text,
  onClick,
  primary = false,
  className,
}: WindowActionButtonProps) => {
  const baseClassName = 'window__action-button';
  const iconOnly = icon && !text;

  return (
    <button
      className={cn(baseClassName, className, {
        [`${baseClassName}--icon-only`]: !!iconOnly,
        button: !iconOnly,
        secondary: !iconOnly && !primary,
      })}
      onClick={onClick}
      type="button"
    >
      {icon && icon}
      {text && text}
    </button>
  );
};

type DefaultCloseButtonProps = {
  iconOnly?: boolean;
  text?: string;
  onClick: () => void;
};

export const DefaultCloseButton = ({
  iconOnly = false,
  text,
  onClick,
  ...otherProps
}: DefaultCloseButtonProps) => {
  const baseClassName = 'window__default-close-button';
  const iconOnlyClass = `${baseClassName}--icon-only`;

  return (
    <WindowActionButton
      className={cn(baseClassName, { [iconOnlyClass]: !!iconOnly })}
      icon={iconOnly ? <CloseIcon width="10" height="10" /> : null}
      text={iconOnly ? undefined : 'Close'}
      primary={!iconOnly}
      onClick={onClick}
      {...otherProps}
    />
  );
};

export default {
  WindowActionButton,
  DefaultCloseButton,
};
