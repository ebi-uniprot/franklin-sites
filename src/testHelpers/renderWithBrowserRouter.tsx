import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

// Inspired by https://testing-library.com/docs/example-react-router/

const renderWithBrowserRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
};

export default renderWithBrowserRouter;
