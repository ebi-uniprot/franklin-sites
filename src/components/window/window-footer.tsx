import { FC } from 'react';

import { DefaultCloseButton } from './window-buttons';

type Props = {
  withCloseButton?: boolean;
  onWindowClose?: () => void;
};

const WindowFooter: FC<Props> = ({
  withCloseButton,
  onWindowClose,
  children,
}) => (
  <div className="window__footer">
    {children}
    {withCloseButton && onWindowClose && (
      <DefaultCloseButton onClick={onWindowClose} text="Close" />
    )}
  </div>
);

export default WindowFooter;
