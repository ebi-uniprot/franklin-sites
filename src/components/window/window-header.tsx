import { DefaultCloseButton } from './window-buttons';

type Props = {
  title: string;
  withCloseButton?: boolean;
  onWindowClose?: () => void;
};

const WindowHeader = ({ title, withCloseButton, onWindowClose }: Props) => (
  <div className="window__header">
    <span className="window__header__title">{title}</span>
    {withCloseButton && onWindowClose && (
      <DefaultCloseButton onClick={onWindowClose} iconOnly />
    )}
  </div>
);

export default WindowHeader;
