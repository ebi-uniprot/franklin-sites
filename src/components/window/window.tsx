import { ReactNode } from 'react';
import cn from 'classnames';

import WindowHeader from './window-header';
import WindowFooter from './window-footer';

import '../../styles/components/window.scss';

type WindowProps = {
  width?: string;
  height?: string;
  title?: string;
  withHeaderCloseButton?: boolean;
  withFooterCloseButton?: boolean;
  onWindowOpen?: () => void;
  onWindowClose?: () => void;
  withShadow?: boolean;
  className?: string;
  actionButtons?: ReactNode;
  children?: ReactNode;
};

const Window = ({
  width = '50vw',
  height = '50vh',
  title,
  withHeaderCloseButton = false,
  withFooterCloseButton = false,
  onWindowOpen,
  onWindowClose,
  actionButtons,
  withShadow = false,
  className,
  children,
}: WindowProps) => {
  const styles = {
    width,
    minHeight: height,
    top: `calc((80vh - ${height}) / 2)`,
    left: `calc((100vw - ${width}) / 2)`,
  };
  const baseClassName = 'window';
  const cssClasses = cn(baseClassName, className, {
    [`${baseClassName}--with-shadow`]: !!withShadow,
  });

  onWindowOpen?.();

  return (
    <div className={cssClasses} style={styles}>
      {title && (
        <WindowHeader
          title={title}
          withCloseButton={withHeaderCloseButton}
          onWindowClose={onWindowClose}
        />
      )}

      <div className="window__content">{children}</div>

      {(withFooterCloseButton || actionButtons) && (
        <WindowFooter
          withCloseButton={withFooterCloseButton}
          onWindowClose={onWindowClose}
        >
          {actionButtons && actionButtons}
        </WindowFooter>
      )}
    </div>
  );
};

export default Window;
