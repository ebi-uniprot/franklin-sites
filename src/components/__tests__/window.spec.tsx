import Window from '../window/window';
import { WindowActionButton } from '../window/window-buttons';
import renderWithBrowserRouter from '../../testHelpers/renderWithBrowserRouter';

const customActionButtons = [
  <WindowActionButton text="No" key="window-action-no" onClick={() => null} />,
  <WindowActionButton
    text="Yes"
    key="window-action-yes"
    onClick={() => null}
    primary
  />,
];

describe('Window component', () => {
  it('should render', () => {
    const { asFragment } = renderWithBrowserRouter(
      <Window
        width="20rem"
        height="15rem"
        title="Alert"
        withHeaderCloseButton
        withFooterCloseButton
        onWindowOpen={() => null}
        onWindowClose={() => null}
        withShadow
        key="full-featured-window"
      >
        Hello, World!
      </Window>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with custom action buttons', () => {
    const { asFragment } = renderWithBrowserRouter(
      <Window
        width="20rem"
        height="10rem"
        title="Confirm"
        actionButtons={customActionButtons}
        key="basic-dialog-window"
      >
        Complete this action?
      </Window>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
