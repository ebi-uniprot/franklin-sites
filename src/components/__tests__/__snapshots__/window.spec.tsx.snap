// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Window component should render 1`] = `
<DocumentFragment>
  <div
    class="window window--with-shadow"
    style="width: 20rem; min-height: 15rem;"
  >
    <div
      class="window__header"
    >
      <span
        class="window__header__title"
      >
        Alert
      </span>
      <button
        class="window__action-button window__default-close-button window__default-close-button--icon-only window__action-button--icon-only"
        type="button"
      >
        <test-file-stub
          height="10"
          width="10"
        />
      </button>
    </div>
    <div
      class="window__content"
    >
      Hello, World!
    </div>
    <div
      class="window__footer"
    >
      <button
        class="window__action-button window__default-close-button button"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Window component should render with custom action buttons 1`] = `
<DocumentFragment>
  <div
    class="window"
    style="width: 20rem; min-height: 10rem;"
  >
    <div
      class="window__header"
    >
      <span
        class="window__header__title"
      >
        Confirm
      </span>
    </div>
    <div
      class="window__content"
    >
      Complete this action?
    </div>
    <div
      class="window__footer"
    >
      <button
        class="window__action-button button secondary"
        type="button"
      >
        No
      </button>
      <button
        class="window__action-button button"
        type="button"
      >
        Yes
      </button>
    </div>
  </div>
</DocumentFragment>
`;
